File: LLD/src/main/java/org/example/accessmodifiers/p2/ChildInDiffPackage.java
package org.example.accessmodifiers.p2;

import org.example.accessmodifiers.p1.Parent;

public class ChildInDiffPackage extends Parent {
    public void checkAccess() {
        System.out.println("\n--- Inside ChildInDiffPackage (Different Package, Child Class) ---");

        System.out.println("publicVar: " + publicVar + " (Accessible)");
        System.out.println("protectedVar: " + protectedVar + " (Accessible via Inheritance)");

        // System.out.println("defaultVar: " + defaultVar);
        System.out.println("defaultVar: NOT Accessible (Default is only visible within the same package)");

        // System.out.println("privateVar: " + privateVar);
        System.out.println("privateVar: NOT Accessible (Private is only visible within the same class)");
    }
}

File: LLD NodeJs/p1_oops/c1_accessmodifiers/p2/ChildInDiffPackage.js
const Parent = require('../p1/Parent');

/**
 * Demonstrates access from a child class in a different module/package
 */
class ChildInDiffPackage extends Parent {
    checkAccess() {
        console.log("\n--- Inside ChildInDiffPackage (Different Package, Child Class) ---");

        console.log("publicVar: " + this.publicVar + " (Accessible)");
        console.log("protectedVar: " + this._protectedVar + " (Accessible via Inheritance - convention)");

        // In JS, defaultVar is still accessible since there's no true package-private
        console.log("defaultVar: " + this.defaultVar + " (Accessible - JS has no package-private)");

        // console.log("privateVar: " + this.#privateVar);
        // SyntaxError: Private field '#privateVar' must be declared in an enclosing class
        console.log("privateVar: NOT Accessible (Private fields with # are only visible within the declaring class)");
    }
}

module.exports = ChildInDiffPackage;
